defrag/defrag.go

```diff
@@ -0,0 +1,276 @@
+	} else {
+		for e := f.List.Front(); e != nil; e = e.Next() {
+			frag, _ := e.Value.(*layers.IPv4)
+			if ip.FragOffset <= frag.FragOffset {
+				debug.Printf("defrag: inserting frag %d before existing frag %d \n",
+					ip.FragOffset*8, frag.FragOffset*8)
+				f.List.InsertBefore(ip, e)
+				break
+			}
+		}
+	}
+	// packet.Metadata().Timestamp should have been better, but
+	// we don't have this info there...
+	f.LastSeen = time.Now()
+
+	fragLength := ip.Length - 20
+	// After inserting the Fragment, we update the counters
+	if f.Total < ip.FragOffset*8+fragLength {
+		f.Total = ip.FragOffset*8 + fragLength
+	}
+	f.Current = f.Current + fragLength
+
+	debug.Printf("defrag: insert ListLen: %d Total:%d Current:%d\n",
+		f.List.Len(),
+		f.Total, f.Current)
+
+	// Final Fragment ?
+	if ip.Flags&layers.IPv4MoreFragments == 0 {
+		f.FinalReceived = true
+	}
+	// Ready to try defrag ?
+	if f.FinalReceived && f.Total == f.Current {
+		return f.Build(ip)
+	}
+	return false, nil
+}
+
+// Build builds the final datagram, modifying ip in place.
+// It puts priority to packet in the early position of the list.
+// See Insert for more details.
+func (f *FragmentList) Build(ip *layers.IPv4) (bool, error) {
+	var final []byte
+	var currentOffset uint16 = 0
+
+	debug.Printf("defrag: building the datagram \n")
+	for e := f.List.Front(); e != nil; e = e.Next() {
+		frag, _ := e.Value.(*layers.IPv4)
+		if frag.FragOffset*8 == currentOffset {
+			debug.Printf("defrag: building - adding %d\n", frag.FragOffset*8)
+			final = append(final, frag.Payload...)
+			currentOffset = currentOffset + frag.Length - 20
+		} else if frag.FragOffset*8 < currentOffset {
+			// overlapping fragment - let's take only what we need
+			startAt := currentOffset - frag.FragOffset*8
+			debug.Printf("defrag: building - overlapping, starting at %d\n",
+				startAt)
+			if startAt > frag.Length-20 {
+				return false, fmt.Errorf("defrag: building - invalid fragment")
+			}
+			final = append(final, frag.Payload[startAt:]...)
+			currentOffset = currentOffset + frag.FragOffset*8
+		} else {
+			// Houston - we have an hole !
+			return false, nil
+		}
+		debug.Printf("defrag: building - next is %d\n", currentOffset)
+	}
+
+	// TODO recompute IP Checksum
+	ip.Payload = final
+	ip.Length = f.Total
+	ip.FragOffset = 0
+	ip.Flags = 0
+
+	return true, nil
+}
+
+// IPv4Flow is a struct to be used as a key.
+type IPv4Flow struct {
+	net gopacket.Flow
+	id  uint16
+}
+
+// NewIPv4Flow returns a new initialized IPv4Flow
+func NewIPv4Flow(ip *layers.IPv4) *IPv4Flow {
+	ipf := new(IPv4Flow)
+	ipf.net = ip.NetworkFlow()
+	ipf.id = ip.Id
+
+	return ipf
+}
+
+// IPv4Defragmenter is a struct which embedded a map of
+// all fragment/packet.
+type IPv4Defragmenter struct {
+	sync.RWMutex
+	IPFlows map[IPv4Flow]*FragmentList
+}
+
+// NewIPv4Defragmenter returns a new IPv4Defragmenter
+// with an initialized map.
+func NewIPv4Defragmenter() *IPv4Defragmenter {
+	d := new(IPv4Defragmenter)
```

Comment:
In fact, the NewIPv4Defrag is intended to be used outside the package, without the need to disclose the internal fields...